cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java

```diff
@@ -225,6 +253,36 @@ public Void apply(AppMds mds) throws Exception {
       }
     });
 
+    if (id.getType() == ProgramType.WORKFLOW && runStatus == ProgramRunStatus.COMPLETED) {
```

Comment:
This piece of code seems to abruptly occur while stopping a program. Let's add some comments to it. Preferably, also move it to its own method.